File: packages/baseline-py/src/baseline_py/gate/run_stage.py
"""Run one stage and classify how it ended."""

import shutil
import subprocess
import time
from pathlib import Path

from baseline_py.gate.build_stage_result import build_stage_result
from baseline_py.gate.stage import Stage
from baseline_py.gate.stage_kind import StageKind
from baseline_py.gate.stage_result import StageResult
from baseline_py.gate.stage_status import StageStatus


def run_stage(stage: Stage, project_root: Path) -> StageResult:
    """Return the stage result.

    A required tool that is not installed is FAILED_TO_RUN, not a skip: a
    gate that silently omits a check and still reports success is worse than
    no gate. A missing shadow tool is different - it is advisory by design
    and optional to install, so its absence is a skip, not an alarm.
    """
    executable = shutil.which(stage.command[0])
    if executable is None:
        if stage.kind is StageKind.SHADOW:
            return build_stage_result(
                stage,
                StageStatus.SKIPPED_NOT_APPLICABLE,
                0,
                0.0,
                "not installed; shadow stages are optional",
            )
        return build_stage_result(
            stage, StageStatus.FAILED_TO_RUN, 127, 0.0, "executable not found"
        )
    started = time.monotonic()
    completed = subprocess.run(
        (executable, *stage.command[1:]),
        cwd=project_root,
        capture_output=True,
        text=True,
        check=False,
    )
    elapsed = time.monotonic() - started
    status = StageStatus.PASSED if completed.returncode == 0 else StageStatus.FINDINGS
    detail = "" if completed.returncode == 0 else completed.stdout.strip()[-2000:]
    return build_stage_result(stage, status, completed.returncode, elapsed, detail)

File: packages/db-quality/src/gate/runStage.ts
import { performance } from 'node:perf_hooks'

import { renderFindings } from '@/check/renderFindings.js'
import type { Stage } from '@/gate/Stage.js'
import type { StageResult } from '@/gate/StageResult.js'
import type { StageStatus } from '@/gate/StageStatus.js'

export const runStage = (stage: Stage): StageResult => {
  const started = performance.now()
  const done = (status: StageStatus, detail: string): StageResult => ({
    name: stage.name,
    status,
    durationSeconds: (performance.now() - started) / 1000,
    detail,
  })
  try {
    const outcome = stage.run()
    if (outcome === 'not-applicable') return done('skipped-not-applicable', '')
    return outcome.length === 0
      ? done('passed', '')
      : done('findings', renderFindings(outcome))
  } catch (error) {
    return done(
      'failed-to-run',
      error instanceof Error ? error.message : String(error),
    )
  }
}
